### ISSUE TITLE
Load macros from file(s)

### ISSUE THREAD

**Issue by Torvaney** [sqlfluff/sqlfluff#156]
Hi folks, I love this project - it is exactly what I have been looking for.

I was wondering if there's a way to load macros from files. I am using dbt and so when I add a macro, I have been adding it both to `dbt/macros/..` as well as `.sqlluff`'s `[sqlfluff:templater:jinja:macros]` field.

Is there any way to load the `[sqlfluff:templater:jinja:macros]` field from files in `dbt/macros`?

**Comment by alanmcruickshank** [comment]
It should be possible and it's a feature I'd like to support.

There are a few bits of this that might be complicated:
- The path to the macros folder might be variable, how would you expect to supply it? Do you think there should be intelligent guesses?
- Should this be a feature you would expect by default, or one that is one you have to specifically activate?
- Can you post an outline of the folder structure that you would expect to run sqlfluff within, pointing out where the the sql files would live and where the macros would live.
- Can you post a couple of example queries and a couple of example macros.
- Would you expect the templated part of the query to be linted in the same way? There's potentially a conflict between well formatted *macros* and well formatted *results* - which do you think should take precedence? At the moment, sqlfluff will expect the *result* to formatted properly, but I'm curious to find out if that's the most useful - or whether it doesn't matter and we should just skip those sections.

**Comment by HassanJaveed84** [comment]
We're also facing the same issue. We're using `dbt` and our macros are defined in a `macros` folder. Some are pretty lengthy, so duplicating them in the config is a pain.

**Comment by alanmcruickshank** [comment]
This should be possible. I don't know if it will be in the next release but I'm up for considering it pretty soon.

**Comment by Torvaney** [comment]
> * The path to the macros folder might be variable, how would you expect to supply it? Do you think there should be intelligent guesses?
> * Should this be a feature you would expect by default, or one that is one you have to specifically activate?

Good questions. I don't have particularly strong opinions. My hunch is that dbt and jinja templating will be used in the minority of cases, in which case, being activated through configuration makes most sense to me. This hunch may be wrong, in which case, defaulting to `./macros/` or checking `./dbt_project.yml` would seem like sane defaults to me.

> * Can you post an outline of the folder structure that you would expect to run sqlfluff within, pointing out where the the sql files would live and where the macros would live.

I have been running sqlfluff with dbt, which lives in `src/dbt`:

```
src/dbt
├── analysis
├── data
├── dbt_modules
├── dbt_project.yml
├── logs
├── macros
├── models
├── packages.yml
├── README.md
├── target
└── tests
```
 I'm much less sure about what the standard structure is, though. I would probably _expect_ to run sqlfluff from within the dbt project (`src/dbt` in this example).

> * Can you post a couple of example queries and a couple of example macros.

Sure!

<details>
  <summary>Macro</summary>

```
{% macro sb_incremental(tbl, source, tbl_id='match_id', source_id='match_id') %}

  {% if is_incremental() %}

    (
      select
        *
      from {{ source }} as s
      where (
        s.{{ source_id }} not in (select distinct {{ tbl_id }} from {{ tbl.name }})
        -- TODO: or match-has-been-updated
      )
    )

  {% else %}

  {{ source }}

  {% endif %}

{% endmacro %}
```

</details>



<details>
  <summary>SQL</summary>
  
```
{{
  config(
    materialized = "incremental",
    unique_key = 'id'
  )
}}


select distinct on (id)
  (json -> 'type' ->> 'id')::int   as id,
  (json -> 'type' ->> 'name')      as name
from {{ sb_incremental(this, 'sb_route_events') }} as e
```

</details>

> * Would you expect the templated part of the query to be linted in the same way? There's potentially a conflict between well formatted macros and well formatted results - which do you think should take precedence? At the moment, sqlfluff will expect the result to formatted properly, but I'm curious to find out if that's the most useful - or whether it doesn't matter and we should just skip those sections.

I think there's a few points here:

* By and large, I'm not so bothered about formatting within the `{{ ... }}` (in templated SQL), since the jinja I have been using is relatively lightweight. I understand that might be different for other people, though.
* Formatted macro definitions would be a definite nice-to-have for me, but not an issue if it's not there
* The priority for me would always be the well-formatted templated SQL, since that's the part that gets read the most.

---

Apologies for the delayed reply. Hope this is useful

**Comment by alanmcruickshank** [comment]
Support for this is coming in a PR shortly, and will be released live this weekend. I'm going to leave this issue open however for feedback (Especially from @Torvaney & @HassanJaveed84).

I'm conscious that I had to make quite a few design decisions in this iteration, some of which I'm not completely sold on yet, so I'd love to hear what the feature is like to work with. Take a look at the tests added in the associated PR for an example of how I imagine this could work.

**Comment by alanmcruickshank** [comment]
The first version of this is live in release [0.3.2](https://github.com/alanmcruickshank/sqlfluff/releases/tag/0.3.2). What do you guys think? I've tested it a bit, but would appreciate some battle testing.

**Comment by HassanJaveed84** [comment]
@alanmcruickshank I'm testing this right now, will get back. 
Thank you for this!

**Comment by pwildenhain** [comment]
Adding my two cents in 💰:

>Would you expect the templated part of the query to be linted in the same way? There's potentially a conflict between well formatted macros and well formatted results - which do you think should take precedence? At the moment, sqlfluff will expect the result to formatted properly, but I'm curious to find out if that's the most useful - or whether it doesn't matter and we should just skip those sections.

I would love the to be able to lint the macros themselves, since people will need to read that if they want to know what your macro is doing

I would also like to be able to flip a switch 💡 and ignore 🙈 lints in any rendered templated SQL sections of the query for two reasons:
1. I _think_ people should generally be reading the templated SQL to understand the query
2. Formatting (read: whitespace control) is really tricky 😵 in `jinja` and I can never seem to get it _perfect_ -- even the `dbt` folks [recommend against](https://docs.getdbt.com/docs/building-a-dbt-project/jinja-macros/#faqs) working to make templated SQL pretty 💇 💄  
    - >My compiled SQL has a lot of spaces and new lines, how can I get rid of it?
        - >This is known as "whitespace control"......Take caution: it's easy to fall down a rabbit hole when it comes to whitespace control!

**Comment by pwildenhain** [comment]
It seems like this is also possibly related to #381, #371, and #335 as far as "getting `sqlfluff` to work with `dbt` macros goes"

**Comment by alanmcruickshank** [comment]
I'm going to suggest that this is best achieved with the new `dbt` templater built into sqlfluff.

For people watching this issue (@noahbruegmann , @dmateusp , @KimchaC , @olhmr ), there is now an alpha release of sqlfluff 0.4.0 available on pypi (https://pypi.org/project/sqlfluff/0.4.0a1/).

I would be very grateful for feedback on the work that @dmateusp has done to get us this far as part of #508 .

In particular you can find docs for the new dbt templater here: https://docs.sqlfluff.com/en/latest/configuration.html#dbt-project-configuration

It would be useful to know whether:
a) the feature satisfies all the requirements here
b) whether the current implementation works in production
b) whether the documentation that we've got is sufficient for users to set things up

**Comment by NiallRees** [comment]
Going to close this issue as I believe it is resolved by 0.4.0. Great to see you here @Torvaney!
